pkgs/tools/audio/alsaequal/default.nix

```diff
@@ -0,0 +1,88 @@
+{ stdenv, fetchurl
+, alsaLib, caps
+, alsaUtils
+}:
+
+stdenv.mkDerivation rec {
+  name = "alsaequal";
+  version = "0.6";
+
+  src = fetchurl {
+    url = "https://thedigitalmachine.net/tools/alsaequal-${version}.tar.bz2";
+    sha256 = "1w3g9q5z3nrn3mwdhaq6zsg0jila8d102dgwgrhj9vfx58apsvli";
+  };
+
+  buildInputs = [ alsaLib ];
+
+  makeFlags = [ "DESTDIR=$(out)" ];
+
+  patches = [ ./makefile.patch ./false_error.patch ./caps_9.x.patch ];
+
+  postPatch = ''
+    sed -i 's#/usr/lib/ladspa/caps\.so#${caps}/lib/ladspa/caps\.so#g' ctl_equal.c pcm_equal.c
+  '';
+
+  preInstall = ''
+    mkdir -p "$out/lib/alsa-lib"
+  '';
+
+  postInstall = ''
+    mkdir -p "$out/etc"
+    cat << EOF >> "$out/etc/asound.conf"
+    ctl_type.equal { lib "$out/lib/alsa-lib/libasound_module_ctl_equal.so" }
+    pcm_type.equal { lib "$out/lib/alsa-lib/libasound_module_pcm_equal.so" }
+
+    ctl.equal {
+      type equal;
+      controls ".cache/alsaequal/alsaequal.bin"
+    }
+
+    pcm.plugequal {
+      type equal;
+      controls ".cache/alsaequal/alsaequal.bin"
+      # Modify the line below if you do not
+      # want to use sound card 0.
+      #slave.pcm "plughw:0,0";
+      # by default we want to play from more sources at time:
+      slave.pcm "plug:dmix";
+    }
+
+    # pcm.equal {
+    # If you do not want the equalizer to be your
+    # default soundcard comment the following
+    # line and uncomment the above line. (You can
+    # choose it as the output device by addressing
+    # it with specific apps,eg mpg123 -a equal 06.Back_In_Black.mp3)
+    pcm.!default {
+      type plug;
+      slave.pcm plugequal;
+    }
+    EOF
+
+    mkdir -p "$out/bin"
+    cat << EOF >> "$out/bin/alsaequal"
+    #!/bin/sh
```

Comment:
This executable is not part of upstream's intention, for as far as I can tell after reading their README and `Makefile`. It seems they intended this to be a library that alsa will load if the alsa's configuration file tells it to. Could you please explain what's the purpose of this script and what it does?

Reply:
For alsa to work with this plugin it needs to know the location of the library, but unfortunately alsa has no dynamic way to add library paths ($ALSA_PLUGIN_DIR doesn't work #6860), so the only viable solution is to statically write library paths in either "/etc/asound.conf" or "\~/.asoundrc". If it was in "/etc/asound.conf", then, as you stated above, it could have been implemented as a module and "/etc/asound.conf" could point to either "${alsaequal}/lib/alsa-lib/..." or to "/run/current-system/sw/lib/alsa-lib/...", but if the user doesn't want to install this library system-wide, then the only option he has is to use "\~/.asoundrc" and make it point to either "${alsaequal}/lib/alsa-lib/..." or "$HOME/nix-profile/lib/alsa-lib/...".

Taking into consideration that (in my opinion) this library is intended to be mainly used with alsamixer to provide real-time adjustable UI for an equalizer, I wrote this script that by default sets up appropriate "\~/.asoundrc" that points to "${alsaequal}/lib/alsa-lib". But as this file is in user-space and location of "${alsaequal}" is prone to change with updates, it checks if "\~/.asoundrc" points to the right location, and if not, makes appropriate changes. After "\~/.asoundrc" is set, it launches "alsamixer -D equal".

Reply:
OK you've somewhat convinced me this package should include in a `$out/etc/asound.conf`. But I'm not sure I see the need for the script. If a user installs this package to their user environment, thry'll be able to find this package's `asound.conf` in `~/.nix-profile/etc` and the library in `~/.nix-profile/lib/alsa-lib`. He can manage his `~/.asoundrc` on his own and point to the library to `~/.nix-profile/lib/alsa-lib` and not `${alsaequal}/lib/alsa-lib`.

Considering this, I think it would be better to introduce changes to https://github.com/NixOS/nixpkgs/blob/0ba1c9f4aff52a49a89fa90888fbefe9aac9bc09/nixos/modules/services/audio/alsa.nix instead.

BTW according to my findings, there is no `include` mechanism in alsa's configuration system but it seems someone has overcome that via Nixpkgs's wrapping:

https://github.com/NixOS/nixpkgs/blob/0ba1c9f4aff52a49a89fa90888fbefe9aac9bc09/pkgs/top-level/all-packages.nix#L16132-L16134

I couldn't find usages of `alsaPluginsWrapper` in Nixpkgs, but I'm positive it was not made in vain.

Reply:
@MarcWeber wrote `alsaPluginsWrapper` in 2009, according to git blame :) Maybe he would be able to explain the original motive behind it.

Reply:
Now that I think about it, it should be just left as a library, despite loosing plug-n-playability, shifting its configuration to the end-user.
If you approve I shall make appropriate changes.

Reply:
I approve removing the shell script `alsaequal`. As for `etc/asound.conf`, it's your choice to make, but as a general rule of thumb, it's considered best practice to stick to upstream's intentions as much as possible :).